File: frameworks/dotnet/minimal-apis/Implementation/Routes/StreamRoutes.cs
using System.Text.Json;

namespace Implementation.Routes;

/// <summary>
/// stream: items.medium's rows written one per line and flushed as each is written, through
/// minimal APIs' stream result. The length is never known, so the answer goes out chunked.
/// </summary>
public static class StreamRoutes
{
    private static readonly byte[] NewLine = "\n"u8.ToArray();

    public static void Map(IEndpointRouteBuilder app)
    {
        app.MapGet("/stream/items", (Payloads p) => Results.Stream(async body =>
        {
            foreach (Item row in p.Medium.Items)
            {
                await JsonSerializer.SerializeAsync(body, row, JsonContext.Default.Item);
                await body.WriteAsync(NewLine);
                await body.FlushAsync();
            }
        }, "application/x-ndjson"));
    }
}

File: frameworks/java/micronaut/Implementation/src/main/java/implementation/routes/StreamRoutes.java
package implementation.routes;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Arrays;

import implementation.Item;
import implementation.Payloads;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.json.JsonMapper;
import org.reactivestreams.Publisher;
import reactor.core.publisher.Flux;

/**
 * stream: items.medium's rows written one per line. The handler returns a Publisher, and Micronaut
 * writes each element it emits as a chunk of the body, so the length is never known and the answer
 * goes out chunked.
 */
@Controller
public class StreamRoutes {

    private static final String NDJSON = "application/x-ndjson";

    private final Payloads p;

    private final JsonMapper json;

    StreamRoutes(Payloads p, JsonMapper json) {
        this.p = p;
        this.json = json;
    }

    @Get(value = "/stream/items", produces = NDJSON)
    public Publisher<byte[]> items() {
        return Flux.fromIterable(p.medium().items()).map(this::line);
    }

    /** The row as JSON, and a newline. */
    private byte[] line(Item row) {
        try {
            byte[] written = json.writeValueAsBytes(row);
            byte[] line = Arrays.copyOf(written, written.length + 1);
            line[written.length] = '\n';
            return line;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
